Zend/zend_compile.c

```diff
@@ -7384,6 +7527,56 @@ void zend_compile_class_decl(znode *result, zend_ast *ast, zend_bool toplevel) /
 		zend_verify_abstract_class(ce);
 	}
 
+	if (ce->ce_flags & ZEND_ACC_ENUM) {
+		zend_verify_enum(ce);
+	}
+
+	if (ce->ce_flags & ZEND_ACC_ENUM) {
+		zval case_default_value;
+		ZVAL_NULL(&case_default_value);
+		zend_string *case_property_name = zend_string_init("case", strlen("case"), 0);
+		zend_type case_type = ZEND_TYPE_INIT_NONE(0);
+		zend_declare_typed_property(ce, case_property_name, &case_default_value, ZEND_ACC_PUBLIC, NULL, case_type);
+		zend_string_release(case_property_name);
+
+		if (ce->enum_primitive_type != IS_UNDEF) {
+			zval value_default_value;
+			ZVAL_NULL(&value_default_value);
+			zend_string *vaue_property_name = zend_string_init("value", strlen("value"), 0);
+			zend_type value_type = ZEND_TYPE_INIT_NONE(0);
+			zend_declare_typed_property(ce, vaue_property_name, &value_default_value, ZEND_ACC_PUBLIC, NULL, value_type);
+			zend_string_release(vaue_property_name);
+		}
+	}
+
+	if (ce->ce_flags & ZEND_ACC_ENUM) {
+		zend_string *cases_func_name = zend_string_init("cases", strlen("cases"), 1);
```

Comment:
Should this be a final method - it may be easier to read code if this was known to be unchanged? Could add a test of intended result of overriding it.

Reply:
Oh right. The method is defined on the user class but we need to make sure the user doesn't define the method themselves, this check might actually be missing.